File: app/api/routers/projects.py
import logging
import uuid
from fastapi import APIRouter, HTTPException, status
from uuid import UUID
from typing import List
from app.schemas.projects import ProjectCreate, ProjectResponse
from app.api.deps import require_supabase

logger = logging.getLogger(__name__)
router = APIRouter()


def _is_uuid(s: str) -> bool:
    try:
        UUID(s)
        return True
    except ValueError:
        return False


@router.get("/", response_model=List[ProjectResponse])
def get_projects():
    try:
        supabase = require_supabase()
        response = supabase.table("projects").select("*").order("created_at", desc=True).execute()
        return [ProjectResponse(**r) for r in (response.data or [])]
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("get_projects failed: %s", e)
        raise HTTPException(status_code=500, detail=str(e)[:200])

@router.post("/", response_model=ProjectResponse, status_code=status.HTTP_201_CREATED)
def create_project(project: ProjectCreate):
    supabase = require_supabase()
    try:
        slug = "proj-" + uuid.uuid4().hex[:8]
        data = {**project.model_dump(exclude_none=True), "slug": slug}
        response = supabase.table("projects").insert(data).execute()
        if not response.data:
            raise HTTPException(status_code=500, detail="Failed to create project")
        return ProjectResponse(**response.data[0])
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Database error: {str(e)}"
        )


@router.get("/{id_or_slug}", response_model=ProjectResponse)
def get_project(id_or_slug: str):
    """Get a single project by UUID or slug. Uses Supabase only."""
    try:
        supabase = require_supabase()
        if _is_uuid(id_or_slug):
            r = supabase.table("projects").select("*").eq("id", id_or_slug).limit(1).execute()
        else:
            r = supabase.table("projects").select("*").eq("slug", id_or_slug).limit(1).execute()
        if r.data and len(r.data) > 0:
            return ProjectResponse(**r.data[0])
        raise HTTPException(status_code=404, detail="Project not found")
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("get_project failed: %s", e)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)[:200]
        )

File: frontend/src/api/projects.ts
const API_BASE = import.meta.env.VITE_API_URL ?? ''

export interface ApiProject {
  id: string
  name: string
  description: string | null
  slug: string | null
  created_at: string
}

export async function getProjects(): Promise<ApiProject[]> {
  const url = API_BASE ? `${API_BASE}/api/projects` : '/api/projects'
  const res = await fetch(url)
  if (!res.ok) {
    if (res.status === 503) throw new Error('Supabase not configured')
    throw new Error(`Failed to fetch projects: ${res.statusText}`)
  }
  return res.json()
}

export async function getProject(idOrSlug: string): Promise<ApiProject> {
  const url = API_BASE ? `${API_BASE}/api/projects/${encodeURIComponent(idOrSlug)}` : `/api/projects/${encodeURIComponent(idOrSlug)}`
  const res = await fetch(url)
  if (!res.ok) {
    if (res.status === 404) throw new Error('Project not found')
    if (res.status === 503) throw new Error('Supabase not configured')
    throw new Error(`Failed to fetch project: ${res.statusText}`)
  }
  return res.json()
}
